concent_api/core/tests/test_unit_custom_manager_queryset.py

```diff
@@ -0,0 +1,161 @@
+            (48, 3, 2, True),
+            (96, 6, 4, True),
+            (192, 12, 8, True),
+            (384, 24, 16, True),
+        ])
+    def test_that_maximum_download_time_query_gives_correct_value(
+            self,
+            minimum_upload_rate,
+            download_leadin_time,
+            concent_messaging_time,
+            custom_protocol_times,
+    ):
+        with override_settings(
+                MINIMUM_UPLOAD_RATE=minimum_upload_rate,
+                DOWNLOAD_LEADIN_TIME=download_leadin_time,
+                CONCENT_MESSAGING_TIME=concent_messaging_time,
+                CUSTOM_PROTOCOL_TIMES=custom_protocol_times,
+        ):
+            maximum_download_deadline = calculate_maximum_download_time(
+                size=self.report_computed_task.size,
+                rate=settings.MINIMUM_UPLOAD_RATE,
+            )
+            assert_that(
+                Subtask.objects_with_timing_columns.get(
+                    subtask_id=self.report_computed_task.task_to_compute.subtask_id).maximum_download_time).is_equal_to(
+                maximum_download_deadline
+            )
+
+    @pytest.mark.django_db
+    def test_that_maximum_download_time_query_gives_correct_value_without_custom_protocol_times(self):
+        with override_settings(
+                CUSTOM_PROTOCOL_TIMES=False,
+        ):
+            maximum_download_deadline = calculate_maximum_download_time(
+                size=self.report_computed_task.size,
+                rate=settings.MINIMUM_UPLOAD_RATE,
+            )
+            assert_that(
+                Subtask.objects_with_timing_columns.get(
+                    subtask_id=self.report_computed_task.task_to_compute.subtask_id).maximum_download_time).is_equal_to(
+                maximum_download_deadline
+            )
+
+    @pytest.mark.django_db
+    @pytest.mark.parametrize(
+        ('minimum_upload_rate', 'download_leadin_time', 'concent_messaging_time', 'custom_protocol_times'), [
+            (48, 3, 2, True),
+            (96, 6, 4, True),
+            (192, 12, 8, True),
+            (384, 24, 16, True),
+        ])
+    def test_that_download_deadline_query_gives_correct_value(
+            self,
+            minimum_upload_rate,
+            download_leadin_time,
+            concent_messaging_time,
+            custom_protocol_times
+    ):
+        with override_settings(
+                MINIMUM_UPLOAD_RATE=minimum_upload_rate,
+                DOWNLOAD_LEADIN_TIME=download_leadin_time,
+                CONCENT_MESSAGING_TIME=concent_messaging_time,
+                CUSTOM_PROTOCOL_TIMES=custom_protocol_times,
+        ):
+            assert_that(
+                Subtask.objects_with_timing_columns.get(
+                    subtask_id=self.report_computed_task.task_to_compute.subtask_id).download_deadline).is_equal_to(
+                (Subtask.objects_with_timing_columns.get(
```

Comment:
This looks really confusing. Parentheses end in the middle of an expression and it's really hard to see which one matches which. You even have a pair of superfluous parentheses (see if you can find it in our version :P):

``` python
assert_that(
    Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id).download_deadline).is_equal_to(
    (Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id).computation_deadline.timestamp() +
    Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id).subtask_verification_time),
)
```
Please wrap it in such a way that it's clear what is nested in what. Properly wrapped it should look like this:

``` python
assert_that(
    Subtask.objects_with_timing_columns.get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).download_deadline
).is_equal_to(
    Subtask.objects_with_timing_columns.get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).computation_deadline.timestamp() +
    Subtask.objects_with_timing_columns.get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).subtask_verification_time
)
```

Or, if you want shorter lines:
``` python
assert_that(
    Subtask.objects_with_timing_columns.
    get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).
    download_deadline
).is_equal_to(
    Subtask.objects_with_timing_columns.
    get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).
    computation_deadline.
    timestamp() +
    Subtask.objects_with_timing_columns.
    get(subtask_id=self.report_computed_task.task_to_compute.subtask_id).
    subtask_verification_time
)
```
or
``` python
assert_that(
    Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id
    ).download_deadline
).is_equal_to(
    Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id
    ).computation_deadline.timestamp() +
    Subtask.objects_with_timing_columns.get(
        subtask_id=self.report_computed_task.task_to_compute.subtask_id
    ).subtask_verification_time
)
```

I personally think that the version with longer ones is the most readable one in this case.